### DISCUSSION TITLE
Create or Update

### DISCUSSION THREAD

**Discussion by donferi** [sensedeep/dynamodb-onetable#110] [Q&A] [answered]
Hello, I'm trying to use Onetable for the first time and having a bit of troubles figuring out how to do this DocumentClient update operation in Onetable:

```js
const Item = await client.update({
    TableName: 'MyTable',
    Key: {
      pk: `USER#donferi`,
      sk: `POSTS`,
    },
    UpdateExpression: `SET #created = if_not_exists(#created, :created), #updated = :updated, #userId = :userId ADD #count :increment`,
    ExpressionAttributeNames: {
      '#count': 'count',
      '#created': 'created',
      '#updated': 'updated',
      '#userId': 'userId',
    },
    ExpressionAttributeValues: {
      ':increment': 1,
      ':created': new Date().toISOString(),
      ':updated': new Date().toISOString(),
      ':userId': donferi,
    },
    ReturnValues: 'ALL_NEW',
  });
};
```

Basically, this a "Create or Increment". First time it will create the document with a `created`, `updated`, `count: 0`.

If you call it again with the same key, it will update the `updated` and `count` will be incremented to `1`.

I've tried using `update` and setting `exists: null`. The issue with this is that it doesn't create a `created` attribute the first time called and starts `count` at `1` instead of `0`.

Would appreciate any advice on this 🙏 

PS: The TS types are a bit off for `exists`, they allow for `boolean | undefined` while the docs say `boolean | null`.

**Comment by mobsense** [comment]
Hi, thanks for your question.

However, I don't understand fully, are you providing UpdateExpression, ExpressionAttributeNames etc INTO OneTable?

But OneTable generates those for you. Can you please post the exact code you are using with Schema definition.

Typically, you use {default: 0} in your schema or set the value to 0 in create.

Then in your update, you can use:

    await User.update({....}, {add: {counter: 1}})

**Reply by donferi** [reply]
Thanks for the quick reply @mobsense! Let me try to explain a bit further.

The code I posted in my question is a contrived example of what I currently have in my app, using vanilla DocumentClient, not OneTable. I'm trying to port it to OneTable tho.

Unfortunately the code is private but here is a hypothetical use-case. Let's say we have this document that will just keep a count of how many posts a user have by incrementing `count`. OneTable Schema:

```js
UserPosts {
  pk: { type: String, value: 'USER#${username}' },
  sk: { type: String, value: 'POSTS' },
  username: { type: String, required: true },
  count: { type: Number, default: 0 },
}
```

With OneTable I can create this by calling something like:

```js
await UserPosts.create({ username: 'donferi' })
```

and like you mention, an update like:

```js
await UserPosts.update({ username: 'donferi' }, { add: { count: 1 } })`
```

However, Dynamodb's update is able to both create or update a document, that's what I'm trying to do here.

I'm trying to figure out if there is a way to call something like this:
```js
const createOrIncrement = (username) => 
  await UserPosts.update(
    { username },
    {
      exists: null,
      add: { count: 1 },
    }
  );
}
```

And for it to act like the vanilla DocumentClient's update which I posted in my question. So basically what I'm trying to do is: if I call `createOrIncrement('donerfi')`, the first time it will create the document and have a `count` of `0`. Then if I call `createOrIncrement('donerfi')` again, it will just increment `count` to `1` and `updated` to the new date.

Does that clarifies a bit the question?

**Comment by mobsense** [comment]
I can't see an easy way to do your conditional update:

    SET #created = if_not_exists(#created, :created)

The create/update works with {exists: null}, but the counter will be set to 1 on the first create.

If you know you are creating, you could change the count value to be 0 or 1 on the first call.

I can't (yet) think of an elegant way to provide conditional update capability like you've expressed. I understand the use-case and need.  Thoughts on a nice API to do that?

**Reply by donferi** [reply]
You are right about the counter being set to `1` on the first call. This is the case with my vanilla `DocumentClient` example too, my mistake. I'm not sure how to do this even with vanilla `DocumentClient`.

Regarding `created`, what if OneTable always sets `created` like this on update if timestamps are enabled?
```
SET #created = if_not_exists(#created, :created)
```

This will allow us to have the `created` attribute for items created using `update` and I don't think it will have any negative consequences (of course, I could be wrong). 


For more complex needs, I can't think of an elegant API since there are just so many ways one could build an update. I was looking at how dynamodb-toolbox does it, and from the [docs](https://github.com/jeremydaly/dynamodb-toolbox#adding-custom-parameters-and-clauses) they allow a third argument:


> For the update method, you can add additional statements to the clauses by specifying arrays as the SET, ADD, REMOVE and DELETE properties. You can also specify additional ExpressionAttributeNames and ExpressionAttributeValues with object values and the system will merge them in with the generated ones.

```js
 let results = await MyEntity.update(item, {}, {
  SET: ['#somefield = :somevalue'],
  ExpressionAttributeNames: { '#somefield': 'somefield' },
  ExpressionAttributeValues: { ':somevalue': 123 }  
})
```

It is not elegant, but it is pretty powerful and let's you build all kind of corner cases.

**Reply by mobsense** [reply]
I've seen that approach. As we dynamically create attribute names, there is potential for clashes and of course, you can't validate what the caller provides in those fields. But I agree, it does provide for lots of corner cases. I'll revisit this and think about it more.

REVISED: actually there is a postFormat hook you can use for that.

```
User.update({...}, {
    postFormat: (model, cmd) => {
        // modify the command as you wish
        return command
    }
```

In your suggestion regarding create, do you mean the timestamp _created or as per your original sample: a created counter?

**Reply by donferi** [reply]
Didn't know about `postFormat`! Looks like a great workaround for the time being.

Re: `create`, I'm referring to the timestamp `_created`, sorry for the confusion.

Basically if we set this for all updates, it will create the timestamp when creating a document using `update`, once created, it shouldn't update the `_created` timestamp anymore.

```
SET #_created = if_not_exists(#_created, :_created)
```

**Comment by mobsense** [comment, marked as answer]
After a bit more code reading, I realize you can already do:

```
    await User.update(properties, {set: {
        name: `if_not_exists(\${name}, {0})`,
    }})
```

However for timestamps, that fails because you cannot have properties in the property set and in the params.set. 

However, however, that is a pretty easy fix to have the set.* properties take precedence. Then you can put updated/created in the set conditionals.

Because the set definitions can contain full "Where" expressions. i.e. ${name} and {value} as well as DynamoDB functions like if_not_exists, you can do exactly what you need -- once we ship the fix to have set.* properties take precedence.

**Reply by donferi** [reply]
Sounds good! When the fix is released, and if you have some time, could you maybe add an example code? I got a bit confused with the syntax of the snippet above. 😅

**Reply by mobsense** [reply]
```
    let now = Date.now()
    user = await User.update({id: user.id, ....}, {
        exists: null,
        set: {
            created: `if_not_exists(\${created}, {${now}})`,
        }
    })
```

NOTE: the ${now} is being expanded by Javascript due to the backticks ``

OneTable sees:

```
{
    created: 'if_not_exists(${created}, {1630371142802})'
```

Assuming you are using Unix epoch dates and not ISO dates.

**Reply by donferi** [reply]
Ahh now I get it, I really appreciate it!

**Reply by mobsense** [reply]
Fix planned for today in 1.7.5

**Reply by donferi** [reply]
Hey @mobsense I think I found an issue, let me know if you need me to create a new issue or if this suffice:

When paired with `add` the set is not being added to the expression.

Modified the test a bit to showcase this, hope it helps.

```js
test('Set conditional', async() => {
    user = await User.update({id: user.id}, {remove: ['status']})
    expect(user.status).toBeUndefined()

    user = await User.update({id: user.id}, {
        add: {
            balance: 1,
        },
        set: {
            status: `if_not_exists(\${status}, {active})`,
        }
    })
    expect(user.balance).toBe(8)
    expect(user.status).toBe('active')
 })
 ```

**Reply by mobsense** [reply]
Thank you. Good pickup. Just committed a fix for that.
